skl/skl.go

```diff
@@ -132,21 +129,25 @@ func decodeValue(value uint64) (valOffset uint32, valSize uint32) {
 
 // NewSkiplist makes a new empty skiplist, with a given arena size.
 func NewSkiplist(arenaSize int64) *Skiplist {
-	buf, err := z.NewBufferWith(int(arenaSize), int(arenaSize), z.UseCalloc)
-	y.Check(err)
-	skl := NewSkiplistWithBuffer(buf, true)
-	return skl
+	buf := z.Calloc(int(arenaSize))
+	s := NewSkiplistWith(buf, true, nil)
+	s.OnClose = func() {
+		z.Free(s.arena.data)
+	}
+	return s
 }
 
-// NewSkiplistWithBuffer makes a new skiplist, with a given buffer.
-func NewSkiplistWithBuffer(buf *z.Buffer, hasVersions bool) *Skiplist {
+// NewSkiplistWith makes a new skiplist, with a given byte slice.
+func NewSkiplistWith(buf []byte, hasVersions bool, grow func(uint32) []byte) *Skiplist {
```

Comment:
Don't pass the grow function. Set it explicitely.

Reply:
We cannot do this because in that case our code would look like this:
```
s := NewSkiplistWith(buf, true)
s.SetGrowFunc(grow)
```
but the function `NewSkiplistWith()` initializes the head node and if we define the grow function after calling `NewSkiplistWith()`, then initializing the head function can fail if the initial buffer is not big enough.